src/replay/consensus.zig

```diff
@@ -0,0 +1,1243 @@
+    slot_tracker.root = new_root;
+    // TODO
+    // - Prune program cache bank_forks.read().unwrap().prune_program_cache(new_root);
+    // - Extra operations as part of setting new root:
+    //   - cleare reward cache root_bank.clear_epoch_rewards_cache
+    //   - extend banks banks.extend(parents.iter());
+    //   - operations around snapshot_controller
+    //   - After setting a new root, prune the banks that are no longer on rooted paths self.prune_non_rooted(root, highest_super_majority_root);
+
+    // Update the progress map.
+    var to_remove = try std.ArrayListUnmanaged(Slot).initCapacity(
+        allocator,
+        progress.map.count(),
+    );
+    defer to_remove.deinit(allocator);
+
+    var it = progress.map.iterator();
+    while (it.next()) |entry| {
+        if (slot_tracker.slots.get(entry.key_ptr.*) == null) {
+            to_remove.appendAssumeCapacity(entry.key_ptr.*);
+        }
+    }
+
+    for (to_remove.items) |key| {
+        _ = progress.map.swapRemove(key);
+    }
+
+    // Update forkchoice
+    try fork_choice.setTreeRoot(&.{
+        .slot = new_root,
+        .hash = root_hash,
+    });
+}
+
+/// Analogous to https://github.com/anza-xyz/agave/blob/234afe489aa20a04a51b810213b945e297ef38c7/core/src/replay_stage.rs#L1029-L1118
+///
+/// Handle fork resets in specific circumstances:
+/// - When a validator needs to switch to a different fork after voting on a fork that becomes invalid
+/// - When a block producer needs to reset their fork choice after detecting a better fork
+/// - When handling cases where the validator's current fork becomes less optimal than an alternative fork
+///
+/// TODO: Currently a placeholder function. Would be implemened when voting and producing blocks is supported.
+fn resetFork(
+    progress: *const ProgressMap,
+    blockstore: *const BlockstoreReader,
+    reset_slot: Slot,
+    last_reset_hash: Hash,
+    last_blockhash: Hash,
+    last_reset_bank_descendants: std.ArrayList(Slot),
+) !void {
+    _ = &progress;
+    _ = &blockstore;
+    _ = &reset_slot;
+    _ = &last_reset_hash;
+    _ = &last_blockhash;
+    _ = &last_reset_bank_descendants;
+}
+
+const testing = std.testing;
+const TreeNode = sig.consensus.fork_choice.TreeNode;
+const TestDB = sig.ledger.tests.TestDB;
+const TestFixture = sig.consensus.replay_tower.TestFixture;
+const MAX_TEST_TREE_LEN = sig.consensus.replay_tower.MAX_TEST_TREE_LEN;
+const Lockout = sig.runtime.program.vote.state.Lockout;
+const createTestReplayTower = sig.consensus.replay_tower.createTestReplayTower;
+const LtHash = sig.core.hash.LtHash;
+
+test "maybeRefreshLastVote - no heaviest slot on same fork" {
+    var prng = std.Random.DefaultPrng.init(91);
+    const random = prng.random();
+
+    const root = SlotAndHash{ .slot = 0, .hash = Hash.initRandom(random) };
+
+    var fixture = try TestFixture.init(testing.allocator, root);
+    defer fixture.deinit(testing.allocator);
+
+    var replay_tower = try createTestReplayTower(
```

Comment:
Missing a `defer replay_tower.deinit(std.testing.allocator)`? Both here and in other tests. Is there a particular reason for why the tests are structured in such a way? I've noticed this a few times.